### DISCUSSION TITLE
Can each wsl2 instance have a different ip?

### DISCUSSION THREAD

**Discussion by dntjr8096** [microsoft/WSL#7463] [Q&A]
Is it possible to set each wsl2 instance to have a different IP?

I tried to configure 1 Master and 3 Worker servers with wsl2, but each instance has the same IP, so I don't know how to solve it.

I want to configure a big data platform environment using wsl2.

It would be possible to assign a hyper-v virtual switch to each instance, but I don't know the detailed method, so please let me know.

**Comment by Biswa96** [comment]
Port forwarding? There is only one WSL2 VM running at once.

**Reply by dntjr8096** [reply]
umm... no port forwarding

I installed a total of 4 servers in centos 7 with wsl2. It consists of one master server and three worker servers, and created them by "wsl --import" with one image.

Then, I access the Master instance and edit the "/etc/sysconfig/network-scripts/ifcfg-eth0" file. I changed dhcp to static and gave a static IP.

Then, when I started the Worker1 server, it was set to the static IP set above.

Can't each instance of wsl2 have a different vnic? I want to configure an internal network that does not use the same IP.

**Comment by pieterhouwen** [comment]
You can hook up your WSL instances to your local network by creating a network bridge between the vEthernet (WSL) adapter and the adapter that is connected to your internal network.

- Go to your control panel
- Go to Network and Sharing Centre
- Click Change adapter options
- Select the vEthernet (WSL) adapter and the other adapter connected to your local network
- Right click -> Create network bridge
- Start your WSL distro
- Open up your /etc/bash.bashrc file or something like that
- Paste:

```
if ! ifconfig | grep 192.168.2.148 1>/dev/null 2>/dev/null; then
sudo ip addr flush dev eth0
sudo ip addr add 192.168.2.148/24 dev eth0
sudo ip link set eth0 up
sudo ip route add default via 192.168.2.1 dev eth0
fi
``` 
Where 192.168.2.148 would be the IP address that you want your distro to have and 192.168.2.1 is your router IP (default gateway).

This will check at startup if the IP address is hooked up to the local network and if not configure it.

# Important 
Windows has the tendency to remove one of the network adapters from the bridged interface upon reboot. So you have to remember to manually add the interfaces to the network bridge by going to the "Change adapter options" screen in the control panel, right-clicking the network bridge and going to Properties and selecting the WSL and local network adapters.

Hope this helps

**Reply by flybyray** [reply]
> ```
> if ! ifconfig | grep 192.168.2.148 1>/dev/null 2>/dev/null; then
> sudo ip addr flush dev eth0
> sudo ip addr add 192.168.2.148/24 dev eth0
> sudo ip link set eth0 up
> sudo ip route add default via 192.168.2.1 dev eth0
> fi
> ```

Is this still valid for wsl2? i see that the above code applied to each instance will setup `eth0` with the last applied one.

> Is it possible to set each wsl2 instance to have a different IP?
Hence this question will not be answered currently. Or is it working for someone?

**Reply by pieterhouwen** [reply]
No I don't think so. Will have to check

**Reply by pieterhouwen** [reply]
Maybe this is relevant for you: https://randombytes.substack.com/p/bridged-networking-under-wsl?s=r

**Comment by totszwai** [comment]
@pieterhouwen Does this still work/apply? I don't see "vEthernet (WSL) adapter" in the Network and Sharing Center anymore. I am on Windows 11 with WSL2.

**Reply by pieterhouwen** [reply]
You're right, I don't see the adapter in my settings on Win11 either. I don't know how to do this on windows 11

**Reply by pieterhouwen** [reply]
You can check out this article: https://randombytes.substack.com/p/bridged-networking-under-wsl?s=r

**Comment by shinzoke** [comment]
Sorry for the necro but euhm, does this still apply to win10? Since my machine ain’t ready for win11 yet 😅

**Reply by marcuszou** [reply]
May not work in WIndows 10 since WSL2 works namespace level. And all of the distro are in the same namespace per WSL architecture.

**Comment by marcuszou** [comment]
It does not work as of Jan 2025.
Honestly it can set the IP address to whatever you want for a single WSL distro;
After you run up the second distro, the script in the second distro will write the new IP to the every distro, because all teh distros share the same namespace. 
Basically it's a no-go for the case of multi-distros using different IP address.

**Reply by marcuszou** [reply]
Another 2 write-ups could help us a bit:
1- https://superuser.com/questions/1715273/wsl2-two-separate-centos-distributions-have-same-eth0-inet-address
2- https://ops.tips/blog/using-network-namespaces-and-bridge-to-isolate-servers/
Article #2 looks a decent approach, more complex tho.:)

**Reply by marcuszou** [reply]
I got 3 WSL2 distros up and tried the scipts in Article #1, and it turns out all of three distros got different IP address, fatastic!
But after I restarted all of the 3 distros, the IP address revert to the one without the running of the Scripts.

This can be resolved by croning a task tho. will try another way and keep yu posted.

**Reply by fkdyufu** [reply]
Any updates on this?

**Comment by ANVLISTENER** [comment]
Wish a lot, to have this luxury in a simple WSL setup. 😢 

Three  Ubunty 24.04 vms for cluster. 

@Biswa96
